sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/CollationTypeCoercion.scala

```diff
@@ -193,88 +144,208 @@ object CollationTypeCoercion {
+    lctOpt match {
+      case Some(lct) =>
+        expressions.map(e => castStringType(e, lct))
+      case _ =>
+        expressions
+    }
   }
 
   /**
-   * Based on the data types of the input expressions this method determines
-   * a collation type which the output will have. This function accepts Seq of
-   * any expressions, but will only be affected by collated StringTypes or
-   * complex DataTypes with collated StringTypes (e.g. ArrayType)
+   * Tries to find the least common StringType among the given expressions.
    */
-  def getOutputCollation(expr: Seq[Expression]): StringType = {
-    val explicitTypes = expr.filter {
-        case _: Collate => true
-        case _ => false
-      }
-      .map(_.dataType.asInstanceOf[StringType].collationId)
-      .distinct
-
-    explicitTypes.size match {
-      // We have 1 explicit collation
-      case 1 => StringType(explicitTypes.head)
-      // Multiple explicit collations occurred
-      case size if size > 1 =>
-        throw QueryCompilationErrors
-          .explicitCollationMismatchError(
-            explicitTypes.map(t => StringType(t))
-          )
-      // Only implicit or default collations present
-      case 0 =>
-        val implicitTypes = expr.filter {
-            case Literal(_, _: StringType) => false
-            case cast: Cast if cast.getTagValue(Cast.USER_SPECIFIED_CAST).isEmpty =>
-              cast.child.dataType.isInstanceOf[StringType]
-            case _ => true
-          }
-          .map(_.dataType)
-          .filter(hasStringType)
-          .map(extractStringType(_).collationId)
-          .distinct
-
-        if (implicitTypes.length > 1) {
-          throw QueryCompilationErrors.implicitCollationMismatchError(
-            implicitTypes.map(t => StringType(t))
-          )
+  private def findLeastCommonStringType(expressions: Seq[Expression]): Option[StringType] = {
+    if (!expressions.exists(e => SchemaUtils.hasNonUTF8BinaryCollation(e.dataType))) {
+      return None
+    }
+
+    val collationContextWinner = expressions.foldLeft(findCollationContext(expressions.head)) {
+      case (Some(left), right) =>
+        findCollationContext(right).flatMap { ctx =>
+          collationPrecedenceWinner(left, ctx)
         }
-        else {
-          implicitTypes.headOption.map(StringType(_)).getOrElse(SQLConf.get.defaultStringType)
+      case (None, _) => return None
+    }
+
+    collationContextWinner.flatMap { cc =>
+      extractStringType(cc.dataType)
+    }
+  }
+
+  /**
+   * Tries to find the collation context for the given expression.
+   * If found, it will also set the [[COLLATION_CONTEXT_TAG]] on the expression,
+   * so that the context can be reused later.
+   */
+  private def findCollationContext(expr: Expression): Option[CollationContext] = {
+
+    def getChildren: Seq[Expression] = expr match {
+      // we don't need to consider the struct name expressions
+      case struct: CreateNamedStruct => struct.valExprs
```

Comment:
I'd like to understand the rationale for handling special expressions in different places:
1. `CollationTypeCoercion.apply`: we need to match different expressions, as each expression defines which inputs to be used to resolve the string collation conflicts
2. here: seems for the same reason? Not all the inputs of an expression should be considered to get the collation.
3. the following pattern match in this method, especially the ` case extract: ExtractValue`: also seems the same reason?

Shall we have a central place to define the list of these special expressions, which returns the string collation related inputs of the expression?

Reply:
Thanks for pointing this out. You are absolutely correct about `CreateNamedStruct` that it shouldn't be here. It should be up in the `apply` with almost identical behavior to the `MapCreate`.

However I don't think `ExtractValue` is in the same boat, as the previous relate to coercing the types of its children, while `ExtractValue` is a special case in calculating the winner collation context of the expression, where we don't want to look at all children like in the general case but instead only look at one child.

So I think it would make sense to have these two places to put specialized expressions in.

Reply:
@cloud-fan Actually it makes more sense to split this into two special cases; `apply` for handling how expression children should be coerced and `getContextRelevantChildren` which returns the children of expression which should be included in the calculation for winning collation context.

Due to `CollationExpressionWalkerSuite` we are decently certain that we are covering cases that we need to and also are protecting ourself in the future once new expressions are added.